### DISCUSSION TITLE
Simplification of Boolean Expressions

### DISCUSSION THREAD

**Discussion by Bobingstern** [josdejong/mathjs#2848] [Ideas]
Is it possible to make a boolean expression simplifier? Maybe just incorporate it into the simplify function and do things like converting $a \ or \ not \ a$ to $1$ etc

**Comment by josdejong** [comment]
That is a nice idea, thanks for sharing Anik.

I suppose `a or not a` should simplify to `true` instead of `1`, right?

Anyone interested in collecting the boolean rules for simplification and adding these rules to `simplify`?

**Reply by Bobingstern** [reply]
Here is a nice table of the simplification [rules](https://www.electronics-tutorials.ws/boolean/boolean-algebra-simplification.html)

A lot of the time you need to do some factoring to simplify it but that should be easy enough

**Comment by Bobingstern** [comment]
You can refer to [this](https://www.boolean-algebra.com/) site for how simplification works

**Reply by josdejong** [reply]
ow that is a nice one 👌

**Comment by Bobingstern** [comment]
is there any better way to this since there can be multiple orientations of certain functions or am i overlooking something:
```js
math.simplify("(not a and b) or (a and not b)", [
  {l:"n and not n", r:"true"},
  "(n and not n1) or (n1 and not n) -> xor(n, n1)",
  "(n and not n1) or (not n and n1) -> xor(n, n1)",
  "(not n and n1) or (n and not n1) -> xor(n, n1)",
  "(not n and n1) or (not n1 and n) -> xor(n, n1)",
])
```

**Reply by josdejong** [reply]
Nice 😎 . 

I think there is also a rule like `not n or not n1 -> n and n1` and the other way around, right?

This is indeed the most straightforward way. There is an alternative way: instead of a string template, you can also provide a JavaScript callback function `function (node) { /* return a new node or the original */ }`. I expect though that in this case it will not lead to more concise code. See docs: https://mathjs.org/docs/reference/functions/simplify.html

**Reply by Bobingstern** [reply]
Thanks, I will try to get all the rules down this way for now. The law you might be referring to is called De Morgan's law. It states `not ( n or n1) -> not n and not n1`

**Comment by Bobingstern** [comment]
How can I implement a rule for factoring? If i have an expression like $a!bc + abc + abd$ but I want to factor to $ac(b + b!) + abd$

**Reply by josdejong** [reply]
You mean like the following rule?

https://github.com/josdejong/mathjs/blob/c51d263d9cc35f17058e2de02d0b32489e96a57d/src/function/algebra/simplify.js#L338-L341

**Reply by Bobingstern** [reply]
yes, I think i figured it out making use of math.simplify.rules
